### PR TITLE
fix: changing catalog env override existing catalog matches

### PR DESCRIPTION
**Pull request by eakmanrq**
Scenario: A user has views being created in a catalog we will call `main` based on the default catalog from the connection (could also come from the model itself). They then add a `environment_catalog_mapping` which maps that models to the same catalog that they were already being mapped to which would mean having the views still be created in `main`. Before this change, we would consider this a catalog change and therefore say that we need to both add and remove those views from the environment. Since add runs before remove this would result in the views being dropped.

We now check the actual view paths and only demote if we see a change. So if the catalog change ends up being no change for the object itself then we won't remove it.

### CODE REVIEW

**Review by izeigerman — APPROVED** [review]

File: sqlmesh/core/state_sync/common.py [hunk 1]
```diff
@@ -195,11 +195,28 @@ def promote(
...
+                    # Can remove the `table_info.fully_qualified_table.catalog is not None` once we make sure that
+                    # `default` schema is included in the fully qualified table and therefore all fully qualified
+                    # tables have a catalog.
+                    if (
+                        table_info.fully_qualified_table.catalog is not None
+                        and table_info.fully_qualified_table.catalog
+                        != environment.catalog_name_override
+                    )
+                    or (
+                        existing_environment
```
Comment (izeigerman) [inline sqlmesh/core/state_sync/common.py:213]: Why not just use existing `environment_catalog_changed`?

Comment (eakmanrq) [reply sqlmesh/core/state_sync/common.py]: I completely rewrote this logic: https://github.com/TobikoData/sqlmesh/pull/2025/commits/7bc8e359310df6594b0a2e5c5990003642d525dd

File: tests/core/test_state_sync.py [hunk 2]
```diff
@@ -465,19 +465,21 @@ def test_promote_snapshots(state_sync: EngineAdapterStateSync, make_snapshot: t.
 
     promotion_result = promote_snapshots(state_sync, [snapshot_a, snapshot_b_old], "prod")
 
-    assert set(promotion_result.added) == set([snapshot_a.table_info, snapshot_b_old.table_info])
+    assert sorted([x.json() for x in promotion_result.added]) == sorted(
```
Comment (tobymao) [inline tests/core/test_state_sync.py:468]: why did you need to change this test?

**Comment by tobymao** [comment]
very nice changes, just one comment on the test changes

File: tests/core/test_state_sync.py (hunk 2 above)
Comment (eakmanrq) [reply tests/core/test_state_sync.py:468]: It seems related to Pydantic version. With Pydantic 1 and I _think_ some Pydantic 2 versions the old version passes but when upgraded to latest it says they are different. So I serialize/sort the results and they correctly determine if they are equal.

Comment (tobymao) [reply tests/core/test_state_sync.py:468]: can you take a look at the actual pydantic error and let’s fix equality? this may patch the test but will still fail if we use eq check in regular code. id rather fix the equality in general as opposed to working around it in tests

Comment (eakmanrq) [reply tests/core/test_state_sync.py:468]: The issue is that Pydantic includes private attributes in the equality check:
https://github.com/pydantic/pydantic/blob/b785d5b1db4eccb922dd2bf17faea5fc842e6a57/pydantic/main.py#L880

We use these for caching values. In this case it is the fully qualified table: https://github.com/TobikoData/sqlmesh/blob/d5fd1b23a8cbb6c0eca1313af65762ebf09c006c/sqlmesh/core/snapshot/definition.py#L371-L375

This is actually a project wide issue that we don't currently address. A bad fix, just to demonstrate the issue, would be this:
```
    def __eq__(self, other):
        return (self.copy().__pydantic_private__ or {}).clear().__eq__((other.copy().__pydantic_private__ or {}).clear()) 
```

The problem obviously being that we would create copies on each equality check. It would be good to sync with @izeigerman on some options since we knows Pydantic the best. I'm fine with fixing this issue being out of scope of this PR.

Comment (tobymao) [reply tests/core/test_state_sync.py:468]: i don't really agree with that given you've already altered the scope of this PR by changing the tests to json(), i think we should fix it, in snapshot for example, we already over equality to simply check the snapshot id, i would not make an equality check that does copy()

Comment (tobymao) [reply tests/core/test_state_sync.py:468]: https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/snapshot/definition.py#L601

Comment (eakmanrq) [reply tests/core/test_state_sync.py:468]: I think we need a general solution that doesn't require making sure to override `__eq__` for each equality check.

Comment (eakmanrq) [reply tests/core/test_state_sync.py:468]: Opening separate PR to try to find a solution everyone is happy with

Comment (eakmanrq) [reply tests/core/test_state_sync.py:468]: Created here: https://github.com/TobikoData/sqlmesh/pull/2035

**Review by izeigerman — APPROVED** [review]
